## Описание проекта

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Знакомство с данными

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('data.csv')

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

Выведим количество пропущенных значений для каждого столбца.

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Один из них — days_employed. Пропуски в этом столбце обработаем на следующем этапе.   
Другой столбец с пропущенными значениями — total_income — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца income_type.   
Например, у человека с типом занятости сотрудник пропуск в столбце total_income должен быть заполнен медианным доходом среди всех записей с тем же типом.

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

В данных могут встречаться артефакты — значения, которые не отражают действительность и появились по какой-то ошибке. Таким артефактом будет отрицательное количество дней трудового стажа в столбце days_employed. Для реальных данных это нормально. Обработаем значения в этом столбце: заменим все отрицательные значения положительными с помощью метода abs().

In [7]:
data['days_employed'] = data['days_employed'].abs()

Для каждого типа занятости выведим медианное значение трудового стажа days_employed в днях.

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

Выведим перечень уникальных значений столбца children.

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

В столбце children есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма data.

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Ещё раз выведим перечень уникальных значений столбца children, чтобы убедиться, что артефакты удалены.

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

Заполним пропуски в столбце days_employed медианными значениями по каждого типа занятости income_type.

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

Заменим вещественный тип данных в столбце total_income на целочисленный с помощью метода astype().

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

Обработаем неявные дубликаты в столбце education. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру.

In [14]:
data['education'] = data['education'].str.lower()

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

Создадим в датафрейме data столбец total_income_category с категориями:  

0–30000 — 'E';  
30001–50000 — 'D';  
50001–200000 — 'C';  
200001–1000000 — 'B';  
1000001 и выше — 'A'.

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Выведим на экран перечень уникальных целей взятия кредита из столбца purpose.

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:  

'операции с автомобилем',  
'операции с недвижимостью',  
'проведение свадьбы',  
'получение образования'.  

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследуйте данные и ответьте на вопросы

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_pivot_1 = data.pivot_table(index=['children'], values = 'debt', aggfunc=['count', 'sum', 'mean'])
data_pivot_1['mean'] = data_pivot_1['mean']*100
data_pivot_1

count   sum      mean
           debt  debt      debt
children                       
0         14091  1063  7.543822
1          4808   444  9.234609
2          2052   194  9.454191
3           330    27  8.181818
4            41     4  9.756098
5             9     0  0.000000

**Вывод:**   
Наибольшее количество людей из выборки находятся в категории бездетных. И в этой категории доля должников наименьшая. Следовательно, с большей долей вероятности мы можем предполагать, что появление детей напрямую коррелирует с платежеспособностью клиентов. Данный вывод можно было сделать руководствуясь здравым смыслом еще до анализа данных,лол.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
data_pivot_2 = data.pivot_table(index=['family_status'], values='debt', aggfunc=['count', 'sum', 'mean'])
data_pivot_2['mean'] = data_pivot_2['mean']*100
data_pivot_2

count  sum      mean
                        debt debt      debt
family_status                              
Не женат / не замужем   2796  273  9.763948
в разводе               1189   84  7.064760
вдовец / вдова           951   63  6.624606
гражданский брак        4134  385  9.313014
женат / замужем        12261  927  7.560558

**Вывод:**  
Анализируя таблицу можно утверждать, что окольцованные люди - лучшие заемщики. Это очевидно из анализа выборки (в данной категории больше всего людей, что не может быть сравнимо, напрмер, с вдовцами и разведенными людьми, хоть у них доля должников и ниже).

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
data_pivot_3 = data.pivot_table(index=['total_income_category'], values='debt', aggfunc=['count', 'sum', 'mean'])
data_pivot_3['mean'] = data_pivot_3['mean']*100
data_pivot_3

count   sum      mean
                        debt  debt      debt
total_income_category                       
A                         25     2  8.000000
B                       5014   354  7.060231
C                      15921  1353  8.498210
D                        349    21  6.017192
E                         22     2  9.090909

**Вывод:**   
Можно предположить, что у людей из категории В лучшие результаты из представленых, так как категорию D можно отсеить из за малой выборки заемщиков.
Худшие же результаты показал народ из катеогрии С. 

### Как разные цели кредита влияют на его возврат в срок?

In [25]:
data_pivot_4 = data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['count', 'sum', 'mean'])
data_pivot_4['mean'] = data_pivot_4['mean']*100
data_pivot_4

count  sum      mean
                           debt debt      debt
purpose_category                              
операции с автомобилем     4279  400  9.347978
операции с недвижимостью  10751  780  7.255139
получение образования      3988  369  9.252758
проведение свадьбы         2313  183  7.911803

**Вывод:**  
Лучшими показателями обладают заемщики с целью, связанной с недвижимостью. А вот больше всего несвоевременные выплат вносят новоиспеченные автомобилисты. 

### Приведите возможные причины появления пропусков в исходных данных.

Существуют 3 ключевых типа пропусков данных. Это MCAR, MAR и MNAR.

### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Лучшее решение - заполнение пропусков медианными значениями по категориям. Если удалить все сторки с пропусками, то это может повлиять на вывод исследования, как и заполнение пропусков "рандомными" данными или средним значением.
Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства.
Для корректного заполнения пропусков необходимо проанализировать выборку значений.

## Вывод

В данной работе было проведено исследование влияния некоторых критериев заемщика на погашение кредита в срок. $$$$
*Заказчик -* кредитный отдел банка. $$$$
*Задача -* выявить влияение семейного положения и количества детей клиента на факт погашения кредита в срок.$$$$
*Входные данные от банка -* статистика о платёжеспособности клиентов. $$$$
При работе с данными были заполнены пропущенные значения, удалены дубликаты и выделены категории для целей кредита и зарплат заемщиков. 
Для оценки результатов были применены сводные таблицы. Выводы по оценке результатов были приведены после расчетов. По результатам исследования можно предположить, что данных для точной оценки кредиторов недостаточно. 

 - Есть ли зависимость между количеством детей и возвратом кредита в срок?$$$$
|Количество детей | Доля должников |
| --- | --- |
|0 | 9.2% |
|1 | 9.5%|
|2 | 8.2% |

   С большей долей вероятности мы можем предполагать, что появление детей напрямую коррелирует с платежеспособностью клиентов.$$$$
 - Есть ли зависимость между семейным положением и возвратом кредита в срок?$$$$
|Доход | Доля должников |
| --- | --- |
|200001–1000000| 7.0%|
|50001–200000| 8.5%|
|30001–50000| 6.0%|

   Люди, находящиеся в браке чаще остальных выплачивают кредит в срок.$$$$
 - Есть ли зависимость между уровнем дохода и возвратом кредита в срок?$$$$
|Цель кредита | Доля должников |
| --- | --- |
|операции с автомобилем| 9.3%|
|операции с недвижимостью| 7.3%|
|получение образования| 9.3%|
|проведение свадьбы| 7.9%|

   Люди с доходом 50001–200000 составляют наибольшую категорию заемщиков и обладают наихудшими результатми. С разрывом в 1.5% на втором месте расположились люди, чей уровень дохода выше на категорию. Берусь предполагать, что особой зависимости своевременного погашения и уровня доода нет.$$$$
 - Как разные цели кредита влияют на его возврат в срок?$$$$
|Семейное положение | Доля должников |
| --- | --- |
|не женат / не замужем|9.8%|
|в разводе|7.1%|
|гражданский брак|9.3%|
|женат / замужем|7.6%|

   Лучшими показателями обладают заемщики с целью, связанной с недвижимостью. А вот больше всего несвоевременные выплат вносят новоиспеченные автомобилисты.$$$$
   
*Рекомендации по сбору данных:*  
Данные лучше поддаются анализу при более большой и равномерной выборке. С помощью таких данных можно делать более точные выводы.
Так же важно отметить, что пропуски и дубликаты в данных могут значительно искажать результаты исследования, поэтому рекомендуется уделять этому особое внимание при сборе.